BrightID/src/actions/apps.ts

```diff
@@ -1,4 +1,129 @@
+import {
+  setApps,
+  addSig,
+  selectAllSigs,
+} from '@/reducer/appsSlice';
+import WISchnorrClient from '@/utils/WISchnorrClient';
+import { strToUint8Array, uInt8ArrayToB64 } from '@/utils/encoding';
+import { NodeApi } from '@/api/brightId';
+
+function isVerified(verifications, verification) {
+  try {
+    const expr = Parser.parse(verification);
+    for (const v of expr.variables()) {
+      if (!verifications[v]) {
+        verifications[v] = false;
+      }
+    }
+    return expr.evaluate(verifications);
+  } catch (err) {
+    console.log(`verification ${verification} can not be evaluated.`, err);
+    return false;
+  }
+}
+
+export const updateBlindSigs = (api) => async (
+  dispatch: dispatch,
+  getState,
+) => {
+  // dispatch(removeAllSigs(getState()));
+  let {
+    user: { verifications, id },
+    keypair: { secretKey },
+    apps: { apps },
+  } = getState();
+
+  const wISchnorrPublics = {};
+  verifications = _.keyBy(verifications, (v) => v.name);
+  const sigs = selectAllSigs(getState());
+  console.log('sigs', sigs);
+  apps = apps.filter((app) => app.usingBlindSig);
+  for (const app of apps) {
+    const vel = app.verificationExpirationLength;
+    const roundedTimestamp = vel ? parseInt(Date.now() / vel) * vel : 0;
+    for (const verification of app.verifications) {
+      const sigInfo = sigs.find(
+        (sig) => sig.app == app.id && sig.verification == verification,
+      );
+      if (sigInfo && sigInfo.roundedTimestamp == roundedTimestamp) {
+        console.log(`sig exists for ${app.name} (${verification})`);
+        continue;
+      }
+      if (!isVerified(verifications, verification)) {
+        console.log(`user is not verified for ${app.name} (${verification})`);
+        continue;
+      }
+
+      try {
+        const network = __DEV__ ? 'test' : 'node';
+        const url = app.nodeUrl || `http://${network}.brightid.org`;
+        const api = new NodeApi({ url, id, secretKey });
+        const { wISchnorrPublic } = await api.getState();
+        if (!wISchnorrPublic) {
+          console.log('wISchnorrPublic is not set');
+          continue;
+        }
+        const client = new WISchnorrClient(wISchnorrPublic);
+        const pub = await api.getPublic(app.id, roundedTimestamp, verification);
+        const uid = CryptoJS.enc.Base64.stringify(
+          CryptoJS.lib.WordArray.random(16),
+        );
+        console.log(`using ${uid} for ${app.name} (${verification}) as uid`);
+        const info = stringify({
+          app: app.id,
+          roundedTimestamp,
+          verification,
+        });
+        const challenge = client.GenerateWISchnorrClientChallenge(
+          pub,
+          info,
+          uid,
+        );
+        console.log(challenge, 'challenge');
+        console.log(info, 'info');
+        const s = stringify({ id, public: pub });
+        const sig = uInt8ArrayToB64(
+          Object.values(nacl.sign.detached(strToUint8Array(s), secretKey)),
```

Comment:
Why the `Object.values()` here? `nacl.sign.detached()` returns an UInt8Array, which is the expected input for `uint8ArrayToB64`.

Reply:
I followed the pattern that is always used on server side like [here](https://github.com/BrightID/BrightID-Node/blob/master/web_services/foxx/brightid/index.js#L326) which is probably wrong.
I think it was working fine because calling `Object.values` on an array return an array without problem. :)
So you're right and you can safely remove that.

Reply:
Okay, i removed the Object.values() call.